File: src/main/java/me/anonymoussoftware/vacancymanager/api/service/FileVacancyService.java
package me.anonymoussoftware.vacancymanager.api.service;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import org.json.JSONArray;
import org.json.JSONObject;
import org.springframework.stereotype.Service;

import lombok.extern.log4j.Log4j;
import me.anonymoussoftware.vacancymanager.api.result.VacancyListResult;
import me.anonymoussoftware.vacancymanager.model.Vacancy;

@Log4j
@Service
public class FileVacancyService {

    public boolean saveVacancies(VacancyListResult vacancies, File file) {
        try (PrintWriter pw = new PrintWriter(new FileOutputStream(file))) {
            JSONObject obj = new JSONObject();
            obj.put("items", vacancies.getVacancies().stream() //
                    .map(v -> v.toJson(false)) //
                    .collect(Collectors.toList()));
            pw.write(obj.toString());
        } catch (FileNotFoundException e) {
            log.error("Error saving vacancies", e);
            return false;
        }
        return true;
    }

    public VacancyListResult openRawVacancies(File file) {
        try (BufferedReader reader = new BufferedReader(
                new InputStreamReader(new FileInputStream(file), StandardCharsets.UTF_8))) {
            String str = reader.lines().collect(Collectors.joining());
            JSONObject obj = new JSONObject(str);
            JSONArray jsonArray = obj.getJSONArray("items");
            List<Vacancy> vacancies = new ArrayList<>();
            for (int i = 0; i < jsonArray.length(); i++) {
                Vacancy vacancy = Vacancy.fromJson(jsonArray.getJSONObject(i));
                vacancies.add(vacancy);
            }
            return new VacancyListResult(vacancies.size(), vacancies);
        } catch (IOException e) {
            log.error("Error opening vacancies", e);
        }
        return null;
    }

}

File: src/main/kotlin/me/anonymoussoftware/vacancymanager/api/service/FileVacancyService.kt
package me.anonymoussoftware.vacancymanager.api.service

import me.anonymoussoftware.vacancymanager.model.Employer
import me.anonymoussoftware.vacancymanager.model.Vacancy
import org.apache.log4j.Logger
import org.json.JSONObject
import org.springframework.stereotype.Service
import java.io.*
import java.nio.charset.StandardCharsets
import java.util.stream.Collectors

@Service
class FileVacancyService {

    fun importVacancies(vacancies: List<Vacancy>, file: File): Boolean {
        try {
            PrintWriter(FileOutputStream(file)).use { pw ->
                val obj = JSONObject()
                obj.put("items", vacancies.stream() //
                    .map { v -> v.toJson(false) } //
                    .collect(Collectors.toList()) as List<JSONObject>)
                pw.write(obj.toString())
            }
        } catch (e: FileNotFoundException) {
            log.error("Error saving vacancies", e)
            return false
        }

        return true
    }

    fun exportVacancies(file: File): List<Vacancy>? {
        try {
            BufferedReader(
                InputStreamReader(FileInputStream(file), StandardCharsets.UTF_8)
            ).use { reader ->
                val str = reader.lines().collect(Collectors.joining())
                val obj = JSONObject(str)
                val jsonArray = obj.getJSONArray("items")
                val vacancies = ArrayList<Vacancy>()
                for (i in 0 until jsonArray.length()) {
                    val vacancy = Vacancy.fromJson(jsonArray.getJSONObject(i))
                    vacancies.add(vacancy)
                }
                return vacancies
            }
        } catch (e: IOException) {
            log.error("Error opening vacancies", e)
        }

        return null
    }

    fun importBannedEmployers(employers: Collection<Employer>, file: File): Boolean {
        try {
            PrintWriter(FileOutputStream(file)).use { pw ->
                val obj = JSONObject()
                obj.put("banned_employers", employers.stream() //
                    .filter { it.isBanned } //
                    .map { v -> v.toJson(true) } //
                    .collect(Collectors.toList()) as List<JSONObject>)
                pw.write(obj.toString())
            }
        } catch (e: FileNotFoundException) {
            log.error("Error saving employers", e)
            return false
        }

        return true
    }

    fun exportBannedEmployers(file: File): List<Employer>? {
        try {
            BufferedReader(
                InputStreamReader(FileInputStream(file), StandardCharsets.UTF_8)
            ).use { reader ->
                val str = reader.lines().collect(Collectors.joining())
                val obj = JSONObject(str)
                val jsonArray = obj.getJSONArray("banned_employers")
                val employers = ArrayList<Employer>()
                for (i in 0 until jsonArray.length()) {
                    val employer = Employer.fromJson(jsonArray.getJSONObject(i))
                    employers.add(employer)
                }
                return employers
            }
        } catch (e: IOException) {
            log.error("Error opening employers", e)
        }

        return null
    }

    companion object {

        private val log = Logger.getLogger(FileVacancyService::class.java)
    }

}
